File: topk-rs/tests/test_query_sort.rs
use test_context::test_context;

use topk_rs::data::literal;
use topk_rs::proto::v1::data::stage::sort_stage::SortOrder;
use topk_rs::query::{field, select};
use topk_rs::Error;

mod utils;
use utils::dataset;
use utils::ProjectTestContext;

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_sort_by_scalar(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([("_id", field("_id"))])
                .sort("published_year")
                .limit(3),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert_eq!(result.len(), 3);
    assert_fields!(&result, ["_id"]);
    assert_doc_ids_ordered!(result, ["harry", "alchemist", "mockingbird"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_sort_by_multiple_fields(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([("_id", field("_id"))])
                .sort([
                    (field("rating"), SortOrder::Desc),
                    (field("published_year"), SortOrder::Asc),
                ])
                .limit(4),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert_doc_ids_ordered!(result, ["moby", "pride", "gatsby", "hobbit"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_sort_by_multiple_with_literal(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([("_id", field("_id"))])
                .sort([
                    (literal(1u32).into(), SortOrder::Desc),
                    (field("published_year"), SortOrder::Asc),
                ])
                .limit(4),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert_doc_ids_ordered!(result, ["pride", "moby", "gatsby", "hobbit"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_sort_by_too_many_fields(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let err = ctx
        .client
        .collection(&collection.name)
        .query(
            select([("_id", field("_id"))])
                .sort(
                    (0..10)
                        .map(|i| (field(format!("field_{i}")), SortOrder::Asc))
                        .collect::<Vec<_>>(),
                )
                .limit(4),
            None,
            None,
        )
        .await
        .expect_err("Query should have failed validation");

    assert!(matches!(
        err,
        Error::InvalidArgument(s) if s.contains("Sort must have at most 8 expressions")
    ));
}

File: topk-py/tests/test_query_sort.py
import pytest
from topk_sdk.query import field, literal, select

from . import ProjectContext
from .utils import dataset, doc_ids_ordered


def test_query_sort_default_asc(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(_id=field("_id")).sort(field("published_year")).limit(3)
    )

    assert doc_ids_ordered(result) == ["pride", "moby", "gatsby"]


@pytest.mark.parametrize(
    "asc,expected",
    [
        (True, ["pride", "moby", "gatsby"]),
        (False, ["harry", "alchemist", "mockingbird"]),
    ],
)
def test_query_sort_asc(ctx: ProjectContext, asc: bool, expected: list[str]):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(_id=field("_id")).sort(field("published_year"), asc).limit(3)
    )

    assert doc_ids_ordered(result) == expected


def test_query_sort_asc_kwarg(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(_id=field("_id")).sort(field("published_year"), asc=False).limit(3)
    )

    assert doc_ids_ordered(result) == ["harry", "alchemist", "mockingbird"]


def test_query_sort_single_expr_list(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(_id=field("_id")).sort([(field("published_year"), "asc")]).limit(3)
    )

    assert doc_ids_ordered(result) == ["pride", "moby", "gatsby"]


def test_query_sort_multiple_exprs(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(_id=field("_id"))
        .sort([(literal(1), "desc"), (field("published_year"), "asc")])
        .limit(4)
    )

    assert doc_ids_ordered(result) == ["pride", "moby", "gatsby", "hobbit"]


def test_query_sort_multiple_exprs_with_asc_arg():
    with pytest.raises(ValueError):
        select(_id=field("_id")).sort([(field("published_year"), "asc")], True)  # type: ignore


def test_query_sort_multiple_exprs_with_asc_kwarg():
    with pytest.raises(ValueError):
        select(_id=field("_id")).sort([(field("published_year"), "asc")], asc=True)  # type: ignore


def test_query_sort_invalid_order():
    with pytest.raises((TypeError, ValueError)):
        select(_id=field("_id")).sort([(field("published_year"), "up")])  # type: ignore


def test_query_sort_missing_order():
    with pytest.raises((TypeError, ValueError)):
        select(_id=field("_id")).sort([field("published_year")])  # type: ignore
